client/packages/invoices/src/OutboundShipment/DetailView/AddFromScannerButton.tsx

```diff
@@ -31,30 +28,42 @@ export const AddFromScannerButtonComponent = ({
   const filterByName = { existsForNameId: { equalTo: otherPartyId } };
   const { hasBarcodeScanner, isScanning, startScanning, stopScan } =
     useBarcodeScannerContext();
-  const { warning } = useNotification();
+  const { error, warning } = useNotification();
 
   const handleScanResult = async (result: ScanResult) => {
     if (!!result.content) {
-      const { content, gtin } = result;
-      const barcode = gtin ?? content;
-      const barcodes = await getBarcodes(barcode);
+      const { content, gtin, batch } = result;
+      const value = gtin ?? content;
+      const barcodes = await getBarcodes(value);
+
+      if (barcodes.totalCount > 0) {
+        const barcode = barcodes.nodes[0];
+        const id = barcode?.itemId;
 
-      if (barcodes.totalCount === 0) {
-        warning(t('error.no-matching-item'))();
-        // TODO: save barcode after selection
-        onAddItem();
-        return;
+        if (!!id) {
+          onAddItem({
+            item: { id },
+            barcode: { ...barcode, batch },
+          } as DraftItem);
```

Comment:
this type of casting is difficult to validate.

One thing that I really liked about Rust from the get go is how much easier it is to review the code knowing that type safety is guaranteed, and there won't be any nulls or undefines. 

The same can be achieved with TypeScript, as long as we don't use `any` or `as` (`as` can be used but we would need to do exhaustive narrowing). If `as` is used, the reviewer would need to make sure that when this type is consumed it has all of the required fields (or risk missing the chance of runtime error), so for this case first have to find where it's used:

AppFromScannerButton(onAddItem) is use by AppBarButtonsComponent(onAddItem), is used by DetailView where onAddItem -> onOpen of useEditModal hook, useEditModal onOpen would set entity which in turn would be parameter passed to OutboundLineEdit draft parameter.

Inside OutboundLineEdit, if draft.item exists it's set as currentItem which is now ItemRowFragment, now have to validate that item is ok to just have the `id` rather then ItemRowFragment. I kind of gave up at this stage, I tried setting DraftItem.item.id to just be { id: string }, and there was a lot of type errors, most of them I think are ok if we change to { id: string } for downstream types but I think some require a full ItemRowFragment (like the placeholder.item). Since I couldn't validate that it's safe to just have the id field on item in DraftLine, I think this area would need extensive manual testing.

It took me about an hour to investigate this `as` statement, I think we would benefit hugely from not using `as` and either limiting fields to just be intersections of the types or using union variants, the benefits we get:
* Much easier to review
* Let type system do the tests for possible runtime errors rather then manual tests or extensive time analysing use case when reviewing or touching this code in the future
* Unions are simpler reason about rather then optional fields

Reply:
Have updated types in 95790d660